pf/internal/schemashim/object_pseudoresource.go

```diff
@@ -119,3 +125,117 @@ func (*objectPseudoResource) Set(key string, value shim.Schema) {
 func (*objectPseudoResource) Delete(key string) {
 	panic("Delete not supported - is it possible to treat this as immutable?")
 }
+
+type tuplePseudoResource struct {
+	schemaOnly
+	attrs map[string]pfutils.Attr
+	tuple pfattr.TypeWithElementTypes
+}
+
+type tupElementAttr struct{ e pfattr.Type }
+
+func (tupElementAttr) GetDeprecationMessage() string  { return "" }
+func (tupElementAttr) GetDescription() string         { return "" }
+func (tupElementAttr) GetMarkdownDescription() string { return "" }
+func (tupElementAttr) IsOptional() bool               { return false }
+func (tupElementAttr) IsRequired() bool               { return true }
+func (tupElementAttr) IsSensitive() bool              { return false }
+func (tupElementAttr) IsComputed() bool               { return false }
+
+func (t tupElementAttr) GetType() attr.Type { return t.e }
+
+func newTuplePseudoResource(t pfattr.TypeWithElementTypes) shim.Resource {
+	attrs := make(map[string]pfutils.Attr, len(t.ElementTypes()))
+	for i, e := range t.ElementTypes() {
+		k := fmt.Sprintf("t%d", i)
+		attrs[k] = pfutils.FromAttrLike(tupElementAttr{e})
+	}
+	return &tuplePseudoResource{
+		schemaOnly: schemaOnly{"tuplePseudoResource"},
+		attrs:      attrs,
+		tuple:      t}
+}
+
+func (*tuplePseudoResource) SchemaVersion() int         { panic("TODO") }
+func (*tuplePseudoResource) DeprecationMessage() string { panic("TODO") }
+
+func (r *tuplePseudoResource) Schema() shim.SchemaMap {
+	return r
+}
+
+func (r *tuplePseudoResource) Get(key string) shim.Schema {
+	v, ok := r.GetOk(key)
+	if !ok {
+		panic(fmt.Sprintf("Missing key: '%s' in tuple with %d elements", key, r.Len()))
+	}
+	return v
+}
+
+func (r *tuplePseudoResource) GetOk(key string) (shim.Schema, bool) {
+	if key == "" || key[0] != 't' {
+		return nil, false
+	}
+	i, err := strconv.Atoi(key[1:])
+	types := r.tuple.ElementTypes()
+	if err != nil || i > len(types) {
+		return nil, false
+	}
+	return newTypeSchema(types[i], r.attrs), true
+}
+
+func (r *tuplePseudoResource) Len() int {
+	return len(r.tuple.ElementTypes())
+}
+
+func (r *tuplePseudoResource) Range(each func(key string, value shim.Schema) bool) {
+	for i, v := range r.tuple.ElementTypes() {
+		k := fmt.Sprintf("t%d", i)
```

Comment:
That looks exactly like objectPseudoResource with `t1,t2,t3` no? Or is tuplePseudoResource has slightly different behavior? Curious, ok either way.

Reply:
It's very similar, but it doesn't involve an `types.ObjectType`. They could probably be unified more cohesively into a `pseudoObjectType[T]` or a type that just holds a map without understanding its origin.

Reply:
Ah so there is a difference. It's the right call then to have a separate struct. Thanks!